lib/ash/type/union.ex

```diff
@@ -377,4 +377,79 @@ defmodule Ash.Type.Union do
       :error
     end
   end
+
+  @impl true
+  def handle_change(nil, %Ash.Union{type: type_name, value: new_value}, constraints) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(nil, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(%Ash.Union{type: type_name, value: old_value}, nil, constraints) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, nil, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(
+        %Ash.Union{type: type_name, value: old_value},
+        %Ash.Union{type: type_name, value: new_value},
+        constraints
+      ) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(
+        %Ash.Union{value: old_value},
+        %Ash.Union{type: type_name, value: new_value},
+        constraints
+      ) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  @impl true
+  # FIXME I don't really understand what to do with other ways this function is
+  # called.
+
+  def prepare_change(_old_value, "", _constraints), do: {:ok, nil}
```

Comment:
We shouldn't need the `""` case here. The type casting will handle if that resolves to a `nil` value according to the type.

Reply:
We need the `nil` case because thats a valid value of all types.